### PR TITLE
Specialize sorting networks for all primitive .NET types

### PR DESCRIPTION
**Pull request by jonathanpeppers**
Add `Sort` overloads for every primitive .NET type: `byte`, `sbyte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `nint`, `nuint`, `char`, `float`, and `double`. Also adds a generic `Sort<T>(T[], IComparer<T>)` for non-primitive types like `string`.

Each primitive type gets:
- `Sort(Span<T>)` and `Sort(T[])` with unrolled network for sizes 27/28
- `Sort` with `IComparer<T>` overloads
- Private `Sort27`/`Sort28` unrolled compare-and-swap methods
- `ArgumentNullException` validation on all array overloads

The generic `Sort<T>` uses the comparer-based network path (not unrolled). C# overload resolution ensures primitives always use their specialized fast overloads.

## Code generation

Updated `generate_unrolled.py` to produce both `NetworkSort.cs` (public API) and `NetworkSort.Unrolled.cs` (unrolled methods) for all 13 types plus the generic overload. Each benchmark is in its own file.

## Tests

1,335 tests covering all primitive types + string, sizes 0-64, stress tests at 27/28, null array/comparer validation for every type.

## Benchmarks

Each type benchmarks both API surfaces fairly: `ArraySort` vs `NetworkSort` (array path), `SpanSort` vs `NetworkSort_Span` (span path).

### Types where NetworkSort wins big (14-16x faster)

| Type | ArraySort (27) | NetworkSort (27) | Speedup |
|---|---|---|---|
| byte | 1,385 ns | 94 ns | **15x** |
| sbyte | 1,409 ns | 98 ns | **14x** |
| short | 1,527 ns | 104 ns | **15x** |
| ushort | 1,394 ns | 97 ns | **14x** |
| float | 1,601 ns | 105 ns | **15x** |
| double | 1,719 ns | 108 ns | **16x** |
| long | 1,460 ns | 102 ns | **14x** |
| nint | 1,474 ns | 102 ns | **14x** |
| nuint | 1,525 ns | 96 ns | **16x** |

### Types where Array.Sort is already SIMD-optimized (~1x)

| Type | ArraySort (27) | NetworkSort (27) | Ratio |
|---|---|---|---|
| int | 103 ns | 98 ns | ~1x |
| uint | 103 ns | 95 ns | ~1x |
| char | 93 ns | 94 ns | ~1x |
| ulong | 114 ns | 101 ns | ~1.1x |

### string (generic path, slower)

| Type | ArraySort (27) | NetworkSort (27) | Ratio |
|---|---|---|---|
| string | 951 ns | 3,606 ns | 0.26x (slower) |

.NET 10 has SIMD-accelerated sort paths for `int`, `uint`, `char`, and `ulong`, so those baselines are already very fast. The generic `Sort<T>` path for `string` uses comparisons through `IComparer<T>`, which is slower than the BCL's internal sort.

NaN handling for float/double tracked in #10, NaN test coverage in #11.

### CODE REVIEW

File: SortingNetworks/NetworkSort.cs [hunk 1]
```diff
@@ -76,4 +345,557 @@ private static void ApplyNetworkWithComparer(Span<int> span, int[] network, ICom
...
+    {
+        int n = span.Length;
+        if (n == 27 || n == 28)
+        {
+            ref float first = ref MemoryMarshal.GetReference(span);
+            if (n == 27)
+                Sort27(ref first);
+            else
+                Sort28(ref first);
+            return;
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply SortingNetworks/NetworkSort.cs:824]: Filed #10 for this.

File: SortingNetworks.Tests/NetworkSortTests.cs [hunk 2]
```diff
@@ -208,6 +180,212 @@ public void Sort_FallsBackForLargeArrays()
...
+    [Theory]
+    [MemberData(nameof(Lengths))]
+    public void Sort_Doubles(int length) =>
+        VerifySort(length, rng => rng.NextDouble() * 2000 - 1000, NetworkSort.Sort);
+
+    [Fact]
+    public void Sort_27Elements_Double() => StressSort(27, rng => rng.NextDouble() * 20000 - 10000, NetworkSort.Sort);
+
+    [Fact]
+    public void Sort_28Elements_Double() => StressSort(28, rng => rng.NextDouble() * 20000 - 10000, NetworkSort.Sort);
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply SortingNetworks.Tests/NetworkSortTests.cs:337]: Filed #11 for this.

File: generate_unrolled.py [hunk 3]
```diff
@@ -74,27 +82,100 @@ def get_network(n):
...
     lines.append(f"    {{")
-    lines.append(f"        int e0 = first;")
+    lines.append(f"        {type_name} e0 = first;")
     for i in range(1, n):
-        lines.append(f"        int e{i} = Unsafe.Add(ref first, {i});")
+        lines.append(f"        {type_name} e{i} = Unsafe.Add(ref first, {i});")
     lines.append(f"")
     for a, b in pairs:
-        lines.append(f"        if (e{a} > e{b}) {{ int temp = e{a}; e{a} = e{b}; e{b} = temp; }}")
+        lines.append(f"        if (e{a} > e{b}) {{ {type_name} temp = e{a}; e{a} = e{b}; e{b} = temp; }}")
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply generate_unrolled.py:96]: Filed #10 for this.

File: SortingNetworks/NetworkSort.cs [hunk 4]
```diff
@@ -76,4 +345,557 @@ private static void ApplyNetworkWithComparer(Span<int> span, int[] network, ICom
...
+    {
+        int n = span.Length;
+        if (n == 27 || n == 28)
+        {
+            ref double first = ref MemoryMarshal.GetReference(span);
+            if (n == 27)
+                Sort27(ref first);
+            else
+                Sort28(ref first);
+            return;
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply SortingNetworks/NetworkSort.cs:896]: Filed #10 for this.

File: SortingNetworks/NetworkSort.cs [hunk 5]
```diff
@@ -11,6 +16,270 @@ namespace SortingNetworks;
...
+    {
+        comparer ??= Comparer<byte>.Default;
+        int n = span.Length;
+        if (n == 27 || n == 28)
+        {
+            ApplyNetworkWithComparer(span, NetworkData.GetNetwork(n), comparer);
+            return;
+        }
+
+        span.Sort(comparer.Compare);
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply SortingNetworks/NetworkSort.cs]: Fixed — changed to `span.Sort(comparer)` in the generator to avoid delegate allocation. All 13 types updated.

File: SortingNetworks/NetworkSort.cs [hunk 6]
```diff
@@ -76,4 +345,557 @@ private static void ApplyNetworkWithComparer(Span<int> span, int[] network, ICom
...
+    }
+
+    /// <summary>
+    /// Sorts an array of <typeparamref name="T"/> using a sorting network when possible.
+    /// </summary>
+    public static void Sort<T>(T[] array, IComparer<T> comparer)
+    {
+        int n = array.Length;
+        if (n == 27 || n == 28)
+        {
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply SortingNetworks/NetworkSort.cs:964]: Fixed — added `ArgumentNullException.ThrowIfNull(array)` and `ArgumentNullException.ThrowIfNull(comparer)` to the generic `Sort<T>` overload.

File: SortingNetworks.Tests/NetworkSortTests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply SortingNetworks.Tests/NetworkSortTests.cs:337]: Tracked in #11 — will add NaN/Infinity/-0 test coverage there.

File: README.md [hunk 7]
```diff
@@ -68,14 +125,24 @@ dotnet test
 dotnet run --project SortingNetworks.Benchmarks -c Release -- --filter *
 ```
 
+## Code generation
+
+The `NetworkSort.cs` and `NetworkSort.Unrolled.cs` files are auto-generated
+by `generate_unrolled.py`. To regenerate after changing the script:
+
+```
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply README.md:133]: Fixed — both README and PR description now say 1,332 tests.

File: generate_unrolled.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply generate_unrolled.py:96]: Tracked in #10 — will emit NaN-aware comparisons for float/double there.

File: SortingNetworks/NetworkSort.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply SortingNetworks/NetworkSort.cs:824]: Tracked in #10.

File: SortingNetworks/NetworkSort.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply SortingNetworks/NetworkSort.cs:896]: Tracked in #10.
